spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,577 @@
+        },
+        environmentUnchanged,
+    }
+
+    action lowerLevelTransportAcceptFailed(node: Address, a: Address): bool = all {
+        require(config.get(node).connections.contains({address: a, state: "incoming"})), 
+        // TODO: perhaps add some checks
+        config' = config.set(node, config.get(node).removeConnectionOutgoing(a)),
+        environmentUnchanged,
+    }
+
+    action lowerLevelLoseConnection(node: Address, c: Connection):bool = all {
+        require(config.get(node).connections.contains(c)),
+        require(config.get(node).connections != Set()),
+        // The following line is obscure magic and we discovered a bug in the cheat sheet
+        config' = config.set(node, config.get(node).loseConnection(c.address)),
+        environmentUnchanged,
+    }
+
+/* 
+ * Upper level actions: callbacks from a reactor or from the operator
+ */
+
+    action reactorStopPeerForError(node: Address, peerID: str):bool = all {
+        config' = config.set(node, config.get(node).stopPeerForError(peerID)),
+        environmentUnchanged,
+    }
+
+    action operatorDialPeers(node:Address, addresses: Set[Address]):bool = all {
+        config' = config.set(node, config.get(node).dialPeers(addresses)),
+        environmentUnchanged,
+    }
+
+/* 
+ * PEX flow
+ */
+
+    // pexRequest and lowerLevelPexAddresses abstract a flow where PEX realizes we need fresh Addresses
+    // calls a seed node, and the seed node returns a set of addresses
+    action pexRequest(node:Address): bool = all {
+        // we do something only of addressbook is a subset of connections
+        require(config.get(node).AddressBook.filter(a => not(config.get(node).connections.exists(c => c.address == a))) == Set()),
+        requestingNodes' = requestingNodes.union(Set(node)),
+        config' = config,
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+    }
+
+    // not sure if the caller of this is "lower lever" or the PEX reactor
+    action lowerLevelPexAddresses(node:Address, addresses: Set[Address]):bool = all {
+        require(requestingNodes.contains(node)),
+        config' = config.set(node, config.get(node).PexAddrs(addresses)),
+        requestingNodes' = requestingNodes.exclude(Set(node)),
+        Environment' = Environment,
+        OperatorAddressBook' = OperatorAddressBook,
+    }
+
+
+    pure def ensurePeers(s: LocalState, addr: Address): LocalState = {
+        if (s.peersOutbound() + s.peersDialing() < s.MaxNumOutboundPeers) {
+            s.dialPeer(addr)
+        }
+        else {
+            s
+        }
+    }
+
+    action pexEnsurePeers(node: Address): bool = all {
+        //val addr = s.AddressBook.chooseSome() // addressbook.pickAddress(Bias)
+        nondet addr = oneOf(config.get(node).AddressBook)
+        config' = config.set(node, config.get(node).ensurePeers(addr)),
+        environmentUnchanged,
+    }
+
```

Comment:
Regarding `pexEnsurePeers` and `ensurePeers`.

In the current logic `ensurePeers` checks the condition (number of outbound peers), then chooses a random address to dial. So `pexEnsurePeers` should call `s.dialPeer(addr)`and be invoked by `ensurePeers`.

Reply:
merge the pure def and the action into an action so that the 'if' is on the right place